packages/apollo-server-core/src/plugin/usageReporting/addTraceRequestStats.ts

```diff
@@ -0,0 +1,428 @@
+        // execution will fail for that field. However, it's possible that the
+        // field is never executed, e.g. due to being in a fragment that's never
+        // executed.
+        //
+        // So we don't consider the request invalid in this case. We instead
+        // consider this particular usage as the field being present and provided
+        // null. This is because:
+        // - Schema changes to remove this input field will make this request
+        //   always fail instead of sometimes (hence why we consider it present).
+        // - Schema changes to change the field's type from nullable to
+        //   non-nullable won't care about these stats, since the field's type
+        //   in this case is non-nullable. At first glance, it might seem like
+        //   we'd want to avoid saying the field is provided null when its type
+        //   is non-nullable because it might be confusing. But this just avoids
+        //   the truth about an ugly edge case in spec. More importantly, with
+        //   this information, we can let users know when their requests are
+        //   doing this, as it may be unintentional on their part.
+        inputFieldIsUsedByRequest({
+          trace,
+          inputObjectTypeName: getNamedType(parentInputType).name,
+          inputFieldName: node.name.value,
+          inputFieldTypeName: inputType.toString(),
+          isPresent: true,
+          isNull: node.value.kind === 'NullValue' || (
+            node.value.kind === 'Variable' &&
+            nullVariableNames.has(node.value.name.value)
+          ),
+          isUndefined: node.value.kind === 'Variable' &&
+            undefinedVariableNames.has(node.value.name.value),
+        });
+      },
+      EnumValue(node: EnumValueNode): void {
+        // The operation has been successfully validated by this stage, so the
+        // non-null assertion here is fine. Look at the Kind.ENUM case in
+        // TypeInfo.enter() in graphql-js for why this works.
+        const inputType = typeInfo.getInputType()!;
+        enumValueIsPresentInRequest({
+          trace,
+          enumTypeName: getNamedType(inputType).name,
+          enumValueName: node.value,
+        });
+      },
+    }));
+  } catch (_) {
+    // At the stage of AS when this is run, variables have not been validated
+    // yet, and accordingly the code that traverses variables may throw. In
+    // those cases, we consider the request itself as invalid. Since the point
+    // of collecting these stats is to understand how schema changes affect
+    // valid operation executions, we collect no stats for invalid requests.
+    trace.perInputTypeStat = Object.create(null);
+    trace.perEnumTypeStat = Object.create(null);
+  }
+}
+
+// This code is adapted from coerceInputValue() in graphql-js. Note that we
+// validate the variables here in addition to collecting stats, as we don't
+// want to collect stats for invalid requests.
+function addTraceInputValueStats({
```

Comment:
>Can we instead change this code to return something along the lines of a list of InputOrEnumStats

I'm hesitant about creating this extra temporary data structure that we'd need to write and read to correctly, it feels like an extra layer of complexity to grok and test (it's also extra CPU and memory, though not significantly so).

>There is a lot of mutability and tracking all of the branching gets really confusing quickly. I think changing the code to return a list that then gets aggregated and updates the trace object at the end of the function would help.

I agree that `Trace` is a very large and mutable object. I've tried replacing `Trace` with:
- `Pick<Trace, 'perEnumTypeStat' | 'perInputTypeStat'>` for `addTraceRequestStats()` and `addTraceInputValueStats()`
- `Pick<Trace, 'perInputTypeStat'>` for `inputFieldIsUsedByRequest()`
- `Pick<Trace, 'perEnumTypeStat'>` for `enumValueIsPresentInRequest()`
to programmatically enforce that we don't modify other parts of the `Trace`, does that help?

>Also it looks like this also handles enum stats? Do we want to break it into a different function for readability sake?

This is one of the confusing parts of the naming in GraphQL spec/`graphql-js`. `addTraceInputValueStats()` is supposed to mimic the naming of `addTraceRequestStats()`; the former means "adds stats from an input value to a trace" while the latter means "add stats from a request to a trace" (really the reason for the "request" part is to clarify we only look at the request in that function and not the response). The phrase "input value" can be confusing; the reason I chose it is because:
- the function in `graphql-js` that this one mimics uses the name `coerceInputValue()` and declares its argument an "input value"
- the actual terminology spec uses for these kinds of things is "input value" (see [this section](https://spec.graphql.org/June2018/#sec-Input-Values)).

So it feels like the right word to use here. The confusion comes in that GraphQL also has "input objects" and that in a GraphQL document the keyword is just `input`, so people will sometimes think of input objects when they see the prefix "input". (Alas, GraphQL also explicitly defines an "input type" [in the spec](https://spec.graphql.org/June2018/#sec-Input-and-Output-Types) as something separate from input objects.) I'm not sure if there's an easy way to get around the naming issue. I've tried everywhere in the code/comments to use the phrase "input object type" and "input object field" when I can, to distinguish from "input type" and "input value" (it's less ambiguous for "input object field" at least if the "object" is left out thankfully).

Reply:
(Might also be worth changing the proto and Druid, though don't know how much work it is to change Druid)